pkg/auth/rbac_test.go

```diff
@@ -1996,3 +2005,185 @@ func Test_handleAuthorizationTypeResponse(t *testing.T) {
+			asm.Mov.Reg(asm.R5, asm.R0),
+		)
+		for i := 0; i < msgSize; i++ {
+			insns = append(insns,
+				asm.LoadMem(asm.R4, asm.RFP, int16(i+1)*-1, asm.Byte),
+				asm.StoreMem(asm.R5, int16(i), asm.R4, asm.Byte),
+			)
+		}
+		insns = append(insns,
+			asm.Mov.Reg(asm.R1, asm.R5),
+			asm.Mov.Imm(asm.R2, flags),
+			asm.FnRingbufSubmit.Call(),
+		)
+	}
+
+	insns = append(insns,
+		asm.Mov.Imm(asm.R0, int32(0)).WithSymbol("exit"),
+		asm.Return(),
+	)
+
+	rbProg, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		License:      "Dual BSD/GPL",
+		Type:         ebpf.XDP,
+		Instructions: insns,
+	})
+	if err != nil {
+		t.Error("Create rbProg failed, err: ", err)
+		rbMap.Close()
+		return nil, nil, err
+	}
+
+	return rbProg, rbMap, nil
+}
+
+func TestRbac_Run(t *testing.T) {
+	type fields struct {
+		policyStore   *policyStore
+		workloadCache cache.WorkloadCache
+	}
+
+	// Common variables in this test func
+	policyStore1 := &policyStore{
+		byKey:       map[string]*security.Authorization{ALLOW_POLICY: policy2_1},
+		byNamespace: byNamespaceAllow,
+	}
+
+	policyStore2 := &policyStore{
+		byKey:       map[string]*security.Authorization{DENY_POLICY: policy2_4},
+		byNamespace: byNamespaceDeny,
+	}
+
+	workloadCache := cache.NewWorkloadCache()
+	workloadCache.AddWorkload(&workloadapi.Workload{
+		Name: "ut-workload",
+		Uid:  "123456",
+		Addresses: [][]byte{
+			{192, 168, 120, 1},
+			{192, 168, 122, 3},
+		},
+		AuthorizationPolicies: []string{ALLOW_AUTH},
+	})
+
+	// Prepare maps
+	sockOpsProg, mapOfTuple, err := genRingbuf(t, MSG_LEN, 0)
+	if err != nil {
+		t.Fatal("Create mapOfTuple failed, err: ", err)
+	}
+	ret, _, err := sockOpsProg.Test(emptyBPFContext)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if errno := syscall.Errno(-int32(ret)); errno != 0 {
+		t.Fatal("Expected 0 as return value, got", errno)
+	}
+	mapOfAuth, err := ebpf.NewMap(&ebpf.MapSpec{
+		Name:       "map_of_auth",
+		Type:       ebpf.Hash,
+		KeySize:    uint32(unsafe.Sizeof(bpfSockTupleV6{})),
+		ValueSize:  uint32(unsafe.Sizeof(uint32(0))),
+		MaxEntries: 4096,
+	})
+	if err != nil {
+		t.Fatal("Create mapOfAuth failed, err: ", err)
+	}
+	t.Cleanup(func() {
+		sockOpsProg.Close()
+		mapOfTuple.Close()
+		mapOfAuth.Close()
+	})
+
+	tests := []struct {
+		name      string
+		fields    fields
+		wantFound bool
+	}{
+		{
+			"1. Allow, records not found in map_of_auth",
+			fields{
+				policyStore:   policyStore1,
+				workloadCache: workloadCache,
+			},
+			false,
+		},
+		{
+			"2. Deny, records found in map_of_auth",
+			fields{
+				policyStore:   policyStore2,
+				workloadCache: workloadCache,
+			},
+			true,
+		},
+	}
+	for _, tt := range tests {
+		r := &Rbac{
+			policyStore:   tt.fields.policyStore,
+			workloadCache: tt.fields.workloadCache,
+		}
+		// Perform auth runner and wait for return
+		ctx, cancelFunc := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(10 * time.Microsecond)
+			cancelFunc()
+		}()
+		r.Run(ctx, mapOfTuple, mapOfAuth)
```

Comment:
I amnot sure how the above make this function return, becase as we can see if the Run runs the reader.Read(), it will be blocking, at the same time, it has no chance to runinto the ctx check case.

Reply:
https://github.com/kmesh-net/kmesh/pull/407/commits/fd1bd5eb7afcbf263e903a13924404bb9c153bf0#diff-259b014bce73c32a01ff7458a12a1359b7245d1f7e99e5af0f9abe7075cef9abR2042
128 same messages are inserted into ringbuf, so reader will keep reading in every loop, which gives time for ctx check case